creating fermionic creation and annihilation operators

khnikhil opened the pull request:
**Description**
Adding spin-1/2 fermionic creation and annihilation operators to close #2165 (for UnitaryHack 2023).

Specifically, in `qutip.core.operators`, I added the `fcreate` and `fdestroy` functions, following the Jordan-Wigner transformation described in [this guide](https://learn.microsoft.com/en-us/azure/quantum/user-guide/libraries/chemistry/concepts/jordan-wigner).

Specifically, I implemented the following using `qutip.tensor` objects, and documented this in the code.

$$a_j = \sigma_z^{\otimes j} \otimes (\frac{\sigma_x + i \sigma_y}{2}) \otimes I^{\otimes N-j-1}$$

$$a_j^{\dagger} = \sigma_z^{\otimes j} \otimes (\frac{\sigma_x - i \sigma_y}{2}) \otimes I^{\otimes N-j-1}$$

Additionally, I wrote a new test in `qutip.tests.core.test_operators` that verified the commutation relations. I also ran the new code through `pycodestyle` to verify it was written nicely.

Please let me know if there are any issues or anything I should be doing on the documentation side! I hope I understood the problem correctly :)

**Related issues or PRs**
close #2165
I closely followed the protocol in #863

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -471,6 +472,90 @@ def create(N, offset=0, *, dtype=None):
+    Parameters
+    ----------
+    N : int
+        Dimension of Hilbert space.
```
This is no longer the dimension of the space, but the number of sites. Maybe one should use a different parameter here instead of `N` to avoid confusion.

khnikhil replied:
good idea, I changed `N` to `n_sites`

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -18,6 +18,7 @@
 
 from . import data as _data
 from .qobj import Qobj
+from .tensor import tensor
```
You could move the importation inside the function to avoid circular importation.

khnikhil replied:
done, thanks!

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -471,6 +472,90 @@ def create(N, offset=0, *, dtype=None):
+    if site == 0:
+        return tensor([destroy(2), *([identity(2)]*(N-1))])
+    else:
+        if N-site-1 > 0:
```
Add a sanity check that `site < N` for both functions.

khnikhil replied:
done, thank you!

BoxiLi commented on qutip/tests/core/test_operators.py:
```diff
@@ -342,3 +344,24 @@ def test_qzero_like(dims, superrep, dtype):
+def test_fcreate_fdestroy(N):
+    identity = qutip.tensor([*([qutip.identity(2)]*(N))])
+    zero_tensor = 0*I
+    for size_0 in range(N):
```
Do you mean `site_0` instead of `size_0`?

khnikhil replied:
yes, thanks

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -471,6 +472,90 @@ def create(N, offset=0, *, dtype=None):
     return qdiags(data, -1, dtype=dtype)
 
 
+def fdestroy(N, site=0):
```
I think `site` can also be a required argument instead of an optional one.

khnikhil replied:
yes, thank you for the comment!

BoxiLi reviewed (changes requested):
Thank you very much for the PR! The implementation and code style look good.

Here are a few comments:
- The two functions share very similar code. Maybe it is better to define one `private` function and then use it in `fcreate` and `fdestroy`.
- You could also add your source to the doc. Please make sure that URL follows the rst syntax used by Sphinx.

khnikhil commented:
@BoxiLi thank you so much for the comments! 

I addressed all your comments in my new commit, but when I try to run the tests with `pytest` I get a strange error:

<img width="570" alt="Screen Shot 2023-05-27 at 1 01 17 PM" src="https://github.com/qutip/qutip/assets/74874354/1b34ab67-a5c0-4223-863d-d7b64109f153">

This seems to be an issue with a circular import in `qutip.core.data.__init__.py` file, but I haven't touched anything in that file, or for that matter, any file besides `qutip.core.operators.py` and `qutip.tests.core.test_operators.py`. I wanted to reach out and ask if this problem has been resolved somewhere else.

Also, to avoid this circular import problem, would it be better if I created a new py file (`qutip.core.fermionic_operators.py`) for these new operators? happy to do this if it would be helpful.

Thank you so much, and please let me know if you have any insight on the `pytest` error :)

BoxiLi commented:
> I addressed all your comments in my new commit, but when I try to run the tests with pytest I get a strange error:

This is likely because qutip is not correctly installed. Could you try uninstalling qutip and reinstalling again?

> Also, to avoid the circular import problem with tensor, would it be better if I created a new py file (qutip.core.fermionic_operators.py) for these new operators? 

A new file for this seems too much for me... I think `operators.py` is the best choice but it includes indeed mostly basic operators not using `tensor`. Maybe there is a better place for this @Ericgig ?

khnikhil commented:
> This is likely because qutip is not correctly installed. Could you try uninstalling qutip and reinstalling again?

@BoxiLi I did this and it worked, thank you! The issue was that I had originally installed it with the `build` package instead of with the setuptools build.

Also, the code now passes both my new tests and existing tests in `qutip.tests.core.test_operators.py`. Additionally, I ran the `pytest qutip/tests -k` command on `f_op`, `fcreate`, and `fdestroy` to run all the tests associated with the new functions, and they pass as well.

Lastly, regarding the circular import problem, I did as you suggested and moved the import inside the `f_op` function.

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -471,6 +471,130 @@ def create(N, offset=0, *, dtype=None):
+    return f_op(n_sites, site, 'creation', dtype=dtype)
+
+
+def f_op(n_sites, site, action, dtype=None):
```
Let's make it a private function `_f_op` and don't add to the public interface `__all__`.

khnikhil replied:
ah I understand, thank you! done

BoxiLi commented on qutip/core/operators.py:
```diff
@@ -471,6 +471,130 @@ def create(N, offset=0, *, dtype=None):
+
+
+def fcreate(n_sites, site, dtype=None):
+    """ Builds a fermionic creation operator.
```
Could you also copy the docstring of `fdestroy` here, with the correct equation?

khnikhil replied:
yes, done

BoxiLi reviewed (changes requested):
Looks good in general. Two minor suggestions on the naming and docstring.

Let's wait a bit for others to give a double-check, if this should be in `operators.py` moved to other places.

khnikhil commented:
hi @BoxiLi, thank you for the minor suggestions! I implemented them and pushed the changes.

> Let's wait a bit for others to give a double-check, if this should be in `operators.py` or moved to other places.

sounds good :)

BoxiLi commented:
@khnikhil I just realized that this function has not been added to the API doc. Could you add `fcreate`, and `fdestroy` to 
`qutip/doc/apidoc/functions.rst`, just like `create`? Sorry that I didn't realize this in the previous check.

Also, could you formulate the equations in math syntax so that Sphinx can render it correctly? For instance, like this:
https://github.com/qutip/qutip/blob/fb7ec127afe0fbfeb4485cbf3ad9ba7b9507fe64/qutip/core/operators.py#L900-L907

Does anyone want to give a double-check? @Ericgig, @hodgestar? Is it okay to put these functions in `operators.py` even if they use `tensor`?

Ericgig reviewed (commented):
I think it's fine to include them in operators.py, unless we create a few more fermionic operators it's not worth it's own file.

Please include the functions in the documentation as Boxi mentioned.

khnikhil commented on qutip/core/operators.py:
```diff
@@ -471,6 +471,138 @@ def create(N, offset=0, *, dtype=None):
+    s_z = 2 * jmat(0.5, 'z', dtype=dtype)
+
+    # sanity check
+    if site >= n_sites:
```
Yes, I added that check as well.

khnikhil commented on qutip/core/operators.py:
```diff
@@ -471,6 +471,138 @@ def create(N, offset=0, *, dtype=None):
+                        either 'creation' or 'destruction.'" % action)
+
+    # build operator
+    if site == 0:
+        return tensor([operator, *([identity(2, dtype=dtype)]*(n_sites-1))])
+    elif n_sites-site-1 > 0:
+        return tensor([*([s_z]*site), operator,
+                       *([identity(2, dtype=dtype)]*(n_sites-site-1))])
+    else:
+        return tensor([*([s_z]*site), operator])
```
Thank you, implemented!

khnikhil commented on qutip/tests/core/test_operators.py:
```diff
@@ -342,3 +344,24 @@ def test_qzero_like(dims, superrep, dtype):
+
+@pytest.mark.parametrize('n_sites', [2, 3, 4, 5])
+def test_fcreate_fdestroy(n_sites):
+    identity = qutip.tensor([*([qutip.identity(2)]*(n_sites))])
+    zero_tensor = 0*identity
```
Thank you, implemented!

khnikhil commented:
hi @BoxiLi and @Ericgig, 

thank you for the comments! I made the following changes:

1. Added `fcreate` and `fdestroy` to `qutip/doc/apidoc/functions.rst`
2. Put the equations in the docstrings into the proper Sphinx `.. math::` format
3. Incorporated Eric's suggestions.

BoxiLi commented:
@khnikhil Something went wrong with the dimension.

For the doctrines, the empty line before and after `..math::` is, unfortunately, part of the syntax and cannot be omitted. So there is a warning and the build fails.

Ericgig commented on qutip/core/operators.py:
```diff
@@ -471,6 +471,142 @@ def create(N, offset=0, *, dtype=None):
+                        either 'creation' or 'destruction.'" % action)
+
+    eye = identity(2, dtype=dtype)
+    opers = [s_z] * site + [operator] + [eye * (n_sites - site - 1)]
```
```suggestion
    opers = [s_z] * site + [operator] + [eye] * (n_sites - site - 1)
```

khnikhil replied:
Thank you!

khnikhil commented:
@BoxiLi @Ericgig thank you for finding these bugs! I ran `test_operators.py` on my local device and it seems to be working now.

BoxiLi commented:
@khnikhil Congrats!
